libvast/vast/project.hpp

```diff
@@ -0,0 +1,277 @@
+    return slice_;
+  }
+
+  // -- container facade -------------------------------------------------------
+
+  [[nodiscard]] table_slice::size_type size() const noexcept {
+    return slice_.rows();
+  }
+
+  [[nodiscard]] iterator begin() const noexcept {
+    if (*this)
+      return {*this, 0};
+    return end();
+  }
+
+  [[nodiscard]] iterator end() const noexcept {
+    return {*this, size()};
+  }
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  /// Construct a table slice projection for a given set of indices (columns).
+  projection(
+    table_slice slice,
+    std::array<table_slice::size_type, sizeof...(Types)> indices) noexcept
+    : slice_{std::move(slice)}, indices_{indices} {
+    // nop
+  }
+
+  // -- concepts ---------------------------------------------------------------
+
+  /// Support CAF's type inspection.
+  template <class Inspector>
+  friend auto inspect(Inspector& f, projection& x) ->
+    typename Inspector::result_type {
+    return f(caf::meta::type_name("vast.projection"), x.slice_, x.indices_);
+  }
+
+  // -- implementation details -------------------------------------------------
+
+private:
+  const table_slice slice_ = {};
+  const std::array<table_slice::size_type, sizeof...(Types)> indices_;
+};
+
+/// Creates a typed view on a given set of columns of a table slice.
+/// @relates projection
+/// @tparam Types... The explicitly specified types of the columns.
+/// @param slice The table slice to project.
+/// @param hints... The hints of the columns, specified as either flat
+/// column indices, offsets, or column names.
+template <class... Types, class... Hints>
+projection<Types...> project(table_slice slice, Hints&&... hints) {
+  static_assert(sizeof...(Types) == sizeof...(Hints),
+                "project requires an equal number of types and hints");
+  const auto& layout = slice.layout();
+  auto find_flat_index_for_hint
+    = [&](const auto& type, auto&& hint) noexcept -> table_slice::size_type {
+    if constexpr (std::is_convertible_v<decltype(hint), offset>) {
+      if (auto field = layout.at(hint))
+        if (congruent(field->type, type))
+          if (auto flat_index = layout.flat_index_at(hint))
+            return *flat_index;
+    } else if constexpr (std::is_constructible_v<std::string_view,
+                                                 decltype(hint)>) {
+      table_slice::size_type flat_index = 0;
+      for (const auto& field : record_type::each{layout}) {
+        const auto full_name = layout.name() + '.' + field.key();
+        auto name_view = std::string_view{full_name};
+        while (true) {
+          if (name_view == hint)
+            if (congruent(field.type(), type))
+              return flat_index;
+          auto colon = name_view.find_first_of('.');
+          if (colon == std::string_view::npos)
+            break;
+          name_view.remove_prefix(colon + 1);
+        }
+        ++flat_index;
+      }
```

Comment:
Create story for streamlining the `record_type` API; this should not be as complicated.

Reply:
Done.